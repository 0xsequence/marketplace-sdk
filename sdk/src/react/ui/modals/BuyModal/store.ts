import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type { Address, Hash } from 'viem';
import type {
	CheckoutOptionsItem,
	MarketplaceKind,
	Step,
} from '../../../_internal';
import { MarketplaceType } from '../../../_internal';

export type CheckoutOptionsSalesContractProps = {
	chainId: number;
	salesContractAddress: Address;
	collectionAddress: Address;
	items: Array<CheckoutOptionsItem>;
	accountAddress: Address;
	customProviderCallback?: (
		onSuccess: (txHash: string) => void,
		onError: (error: Error) => void,
		onClose: () => void,
	) => void;
};

export type PaymentModalProps = {
	collectibleId: string;
	marketplace: MarketplaceKind;
	orderId: string;
	customCreditCardProviderCallback?: (buyStep: Step) => void;
};

export type BuyModalBaseProps = {
	chainId: number;
	collectionAddress: Address;
	skipNativeBalanceCheck?: boolean;
	nativeTokenAddress?: Address;
	marketplaceType: MarketplaceType;
};

// Shop type modal props
export type ShopBuyModalProps = BuyModalBaseProps & {
	marketplaceType: MarketplaceType.SHOP;
	salesContractAddress: Address;
	items: Array<CheckoutOptionsItem>;
	customProviderCallback?: CheckoutOptionsSalesContractProps['customProviderCallback'];
};

// Marketplace type modal props
export type MarketplaceBuyModalProps = BuyModalBaseProps & {
	marketplaceType: MarketplaceType.MARKET;
	collectibleId: string;
	marketplace: MarketplaceKind;
	orderId: string;
	customCreditCardProviderCallback?: PaymentModalProps['customCreditCardProviderCallback'];
};

// Union type for either shop or marketplace
export type BuyModalProps = ShopBuyModalProps | MarketplaceBuyModalProps;

// Type guard functions
export function isShopProps(props: BuyModalProps): props is ShopBuyModalProps {
	return props.marketplaceType === MarketplaceType.SHOP;
}

export function isMarketplaceProps(
	props: BuyModalProps,
): props is MarketplaceBuyModalProps {
	return props.marketplaceType === MarketplaceType.MARKET;
}

export type onSuccessCallback = ({
	hash,
	orderId,
}: {
	hash?: Hash;
	orderId?: string;
}) => void;
export type onErrorCallback = (error: Error) => void;

const initialContext = {
	isOpen: false,
	props: null as unknown as BuyModalProps,
	onError: (() => {}) as onErrorCallback,
	onSuccess: (() => {}) as onSuccessCallback,
	quantity: undefined as number | undefined,
};

export const buyModalStore = createStore({
	context: { ...initialContext },
	on: {
		open: (
			context,
			event: {
				props: BuyModalProps;
				onError?: onErrorCallback;
				onSuccess?: onSuccessCallback;
			},
		) => ({
			...context,
			props: event.props,
			onError: event.onError ?? context.onError,
			onSuccess: event.onSuccess ?? context.onSuccess,
			isOpen: true,
		}),

		close: (context) => ({
			...context,
			isOpen: false,
			quantity: undefined,
		}),

		setQuantity: (context, event: { quantity: number }) => ({
			...context,
			quantity: event.quantity,
		}),
	},
});

export const useIsOpen = () =>
	useSelector(buyModalStore, (state) => state.context.isOpen);

export const useBuyModalProps = () =>
	useSelector(buyModalStore, (state) => state.context.props);

export const useOnError = () =>
	useSelector(buyModalStore, (state) => state.context.onError);

export const useOnSuccess = () =>
	useSelector(buyModalStore, (state) => state.context.onSuccess);

export const useQuantity = () =>
	useSelector(buyModalStore, (state) => state.context.quantity);
