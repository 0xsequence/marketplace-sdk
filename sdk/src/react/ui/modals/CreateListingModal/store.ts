import { observable } from '@legendapp/state';
import { addDays } from 'date-fns/addDays';
import type { Hex } from 'viem';
import { type Currency, OrderbookKind } from '../../../../types';
import type { CollectionType, TransactionSteps } from '../../../_internal';
import type { BaseModalState, ModalCallbacks } from '../_internal/types';

type CreateListingState = BaseModalState & {
	collectibleId: string;
	collectionName: string;
	orderbookKind?: OrderbookKind;
	collectionType: CollectionType | undefined;
	listingPrice: {
		amountRaw: string;
		currency: Currency;
	};
	quantity: string;
	invalidQuantity: boolean;
	expiry: Date;
	steps: TransactionSteps;
	listingIsBeingProcessed: boolean;
};

export type OpenCreateListingModalArgs = {
	collectionAddress: Hex;
	chainId: number;
	collectibleId: string;
	orderbookKind?: OrderbookKind;
	callbacks?: ModalCallbacks;
};

type Actions = {
	open: (args: OpenCreateListingModalArgs) => void;
	close: () => void;
};

const listingPrice = {
	amountRaw: '0',
	currency: {} as Currency,
};

const approval = {
	exist: false,
	isExecuting: false,
	execute: () => Promise.resolve(),
};

const transaction = {
	exist: false,
	isExecuting: false,
	execute: () => Promise.resolve(),
};

const steps = {
	approval: { ...approval },
	transaction: { ...transaction },
};

const initialState: CreateListingState = {
	isOpen: false,
	collectionAddress: '' as Hex,
	chainId: 0,
	collectibleId: '',
	orderbookKind: OrderbookKind.sequence_marketplace_v2,
	collectionName: '',
	collectionType: undefined,
	listingPrice: { ...listingPrice },
	quantity: '1',
	invalidQuantity: false,
	expiry: new Date(addDays(new Date(), 7).toJSON()),
	callbacks: undefined as ModalCallbacks | undefined,
	steps: { ...steps },
	listingIsBeingProcessed: false,
};

const actions: Actions = {
	open: (args) => {
		createListingModal$.collectionAddress.set(args.collectionAddress);
		createListingModal$.chainId.set(args.chainId);
		createListingModal$.collectibleId.set(args.collectibleId);
		createListingModal$.orderbookKind.set(args.orderbookKind);
		createListingModal$.callbacks.set(args.callbacks);
		createListingModal$.isOpen.set(true);
	},
	close: () => {
		createListingModal$.isOpen.set(false);
		createListingModal$.set({ ...initialState, ...actions });
		createListingModal$.listingPrice.set({ ...listingPrice });
		createListingModal$.steps.set({ ...steps });
		createListingModal$.listingIsBeingProcessed.set(false);
		createListingModal$.steps.approval.isExecuting.set(false);
		createListingModal$.steps.transaction.isExecuting.set(false);
	},
};

export const createListingModal$ = observable<CreateListingState & Actions>({
	...initialState,
	...actions,
});
