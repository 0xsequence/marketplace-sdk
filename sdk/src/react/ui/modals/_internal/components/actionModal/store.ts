import { type Observable, observable } from '@legendapp/state';
import type { Address } from 'viem';

export interface ActionModalState {
	isOpen: boolean;
	chainId: number | null;
	collectionAddress: Address | null;
}

export function createActionModalStore() {
	return observable<ActionModalState>({
		isOpen: false,
		chainId: null,
		collectionAddress: null,
	});
}

export function openModal(store: Observable<ActionModalState>) {
	store.isOpen.set(true);
}

export function closeModal(store: Observable<ActionModalState>) {
	store.isOpen.set(false);
}
