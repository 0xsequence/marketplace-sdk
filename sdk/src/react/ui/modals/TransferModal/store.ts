import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type { Hex } from 'viem';
import type { CollectionType } from '../../../_internal';
import type { ModalCallbacks } from '../_internal/types';
import type { ShowTransferModalArgs } from '.';

export type TransferModalView =
	| 'enterReceiverAddress'
	| 'followWalletInstructions'
	| undefined;

export interface TransferModalState {
	isOpen: boolean;
	state: {
		chainId: number;
		collectionAddress: Hex;
		collectionType?: CollectionType | undefined;
		collectibleId: string;
		quantity: string;
		receiverAddress: string;
		callbacks?: ModalCallbacks;
		transferIsBeingProcessed: boolean;
	};
	view: TransferModalView;
	hash: Hex | undefined;
}

const initialContext: TransferModalState = {
	isOpen: false,
	state: {
		receiverAddress: '',
		collectionAddress: '0x',
		chainId: 0,
		collectibleId: '',
		quantity: '1',
		transferIsBeingProcessed: false,
	},
	view: 'enterReceiverAddress',
	hash: undefined,
};

export const transferModalStore = createStore({
	context: initialContext,
	on: {
		open: (context, event: ShowTransferModalArgs) => ({
			...context,
			isOpen: true,
			state: {
				...context.state,
				chainId: event.chainId,
				collectionAddress: event.collectionAddress,
				collectibleId: event.collectibleId,
				callbacks: event.callbacks,
			},
		}),
		close: () => ({
			...initialContext,
		}),
		setView: (context, event: { view: TransferModalView }) => ({
			...context,
			view: event.view,
		}),
		setHash: (context, event: { hash: Hex | undefined }) => ({
			...context,
			hash: event.hash,
		}),
		setReceiverAddress: (context, event: { address: string }) => ({
			...context,
			state: {
				...context.state,
				receiverAddress: event.address,
			},
		}),
		setQuantity: (context, event: { quantity: string }) => ({
			...context,
			state: {
				...context.state,
				quantity: event.quantity,
			},
		}),
		setTransferIsBeingProcessed: (
			context,
			event: { isProcessing: boolean },
		) => ({
			...context,
			state: {
				...context.state,
				transferIsBeingProcessed: event.isProcessing,
			},
		}),
		updateState: (context, event: Partial<TransferModalState['state']>) => ({
			...context,
			state: {
				...context.state,
				...event,
			},
		}),
	},
});

// Selector hooks
export const useIsOpen = () =>
	useSelector(transferModalStore, (state) => state.context.isOpen);

export const useModalState = () =>
	useSelector(transferModalStore, (state) => state.context.state);

export const useView = () =>
	useSelector(transferModalStore, (state) => state.context.view);

export const useHash = () =>
	useSelector(transferModalStore, (state) => state.context.hash);
