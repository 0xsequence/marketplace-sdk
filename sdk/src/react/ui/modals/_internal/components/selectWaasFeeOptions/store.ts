'use client';

import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type {
	FeeOption,
	WaasFeeOptionConfirmation,
} from '../../../../../../types/waas-types';

interface SelectWaasFeeOptionsContext {
	selectedFeeOption: FeeOption | undefined;
	pendingFeeOptionConfirmation: WaasFeeOptionConfirmation | undefined;
	isVisible: boolean;
}

export const selectWaasFeeOptionsStore = createStore({
	context: {
		selectedFeeOption: undefined,
		pendingFeeOptionConfirmation: undefined,
		isVisible: false,
	} as SelectWaasFeeOptionsContext,
	on: {
		show: (context) => ({
			...context,
			isVisible: true,
		}),
		hide: () => ({
			selectedFeeOption: undefined,
			pendingFeeOptionConfirmation: undefined,
			isVisible: false,
		}),
		setSelectedFeeOption: (
			context,
			event: { feeOption: FeeOption | undefined },
		) => ({
			...context,
			selectedFeeOption: event.feeOption,
		}),
		setPendingFeeOptionConfirmation: (
			context,
			event: { confirmation: WaasFeeOptionConfirmation | undefined },
		) => ({
			...context,
			pendingFeeOptionConfirmation: event.confirmation,
		}),
	},
});

// React hooks
export const useSelectWaasFeeOptionsStore = () => {
	const isVisible = useSelector(
		selectWaasFeeOptionsStore,
		(state) => state.context.isVisible,
	);
	const selectedFeeOption = useSelector(
		selectWaasFeeOptionsStore,
		(state) => state.context.selectedFeeOption,
	);
	const pendingFeeOptionConfirmation = useSelector(
		selectWaasFeeOptionsStore,
		(state) => state.context.pendingFeeOptionConfirmation,
	);

	return {
		isVisible,
		selectedFeeOption,
		pendingFeeOptionConfirmation,
		show: () => selectWaasFeeOptionsStore.send({ type: 'show' }),
		hide: () => selectWaasFeeOptionsStore.send({ type: 'hide' }),
		setSelectedFeeOption: (feeOption: FeeOption | undefined) =>
			selectWaasFeeOptionsStore.send({
				type: 'setSelectedFeeOption',
				feeOption,
			}),
		setPendingFeeOptionConfirmation: (
			confirmation: WaasFeeOptionConfirmation | undefined,
		) =>
			selectWaasFeeOptionsStore.send({
				type: 'setPendingFeeOptionConfirmation',
				confirmation,
			}),
	};
};
