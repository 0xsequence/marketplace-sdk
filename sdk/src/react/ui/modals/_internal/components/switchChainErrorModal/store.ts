import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type { ShowSwitchChainModalArgs } from '.';

export interface SwitchChainErrorModalState {
	isOpen: boolean;
	chainIdToSwitchTo: number | undefined;
	isSwitching: boolean;
}

const initialContext: SwitchChainErrorModalState = {
	isOpen: false,
	chainIdToSwitchTo: undefined,
	isSwitching: false,
};

export const switchChainErrorModalStore = createStore({
	context: initialContext,
	on: {
		open: (context, event: ShowSwitchChainModalArgs) => ({
			...context,
			isOpen: true,
			chainIdToSwitchTo: event.chainIdToSwitchTo,
		}),
		close: (context) => ({
			...context,
			isOpen: false,
			chainIdToSwitchTo: undefined,
			isSwitching: false,
		}),
	},
});

// Selector hooks
export const useIsOpen = () =>
	useSelector(switchChainErrorModalStore, (state) => state.context.isOpen);

export const useChainIdToSwitchTo = () =>
	useSelector(
		switchChainErrorModalStore,
		(state) => state.context.chainIdToSwitchTo,
	);
