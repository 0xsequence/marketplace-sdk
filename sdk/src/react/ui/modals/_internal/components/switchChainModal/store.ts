import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type { SwitchChainError } from 'viem';
import type { ShowSwitchChainModalArgs } from '.';

export interface SwitchChainModalState {
	isOpen: boolean;
	chainIdToSwitchTo: number | undefined;
	isSwitching: boolean;
	onSuccess: (() => void) | undefined;
	onError: undefined | ((error: SwitchChainError) => void);
	onClose: (() => void) | undefined;
}

const initialContext: SwitchChainModalState = {
	isOpen: false,
	chainIdToSwitchTo: undefined,
	isSwitching: false,
	onSuccess: undefined,
	onError: undefined,
	onClose: undefined,
};

export const switchChainModalStore = createStore({
	context: initialContext,
	on: {
		open: (context, event: ShowSwitchChainModalArgs) => ({
			...context,
			isOpen: true,
			chainIdToSwitchTo: event.chainIdToSwitchTo,
			onSuccess: event.onSuccess,
			onError: event.onError,
			onClose: event.onClose,
		}),
		close: (context) => ({
			...context,
			isOpen: false,
			chainIdToSwitchTo: undefined,
			isSwitching: false,
			onSuccess: undefined,
			onError: undefined,
			onClose: undefined,
		}),
		setSwitching: (context, event: { isSwitching: boolean }) => ({
			...context,
			isSwitching: event.isSwitching,
		}),
	},
});

// Selector hooks
export const useIsOpen = () =>
	useSelector(switchChainModalStore, (state) => state.context.isOpen);

export const useChainIdToSwitchTo = () =>
	useSelector(
		switchChainModalStore,
		(state) => state.context.chainIdToSwitchTo,
	);

export const useIsSwitching = () =>
	useSelector(switchChainModalStore, (state) => state.context.isSwitching);

export const useOnSuccess = () =>
	useSelector(switchChainModalStore, (state) => state.context.onSuccess);

export const useOnError = () =>
	useSelector(switchChainModalStore, (state) => state.context.onError);

export const useOnClose = () =>
	useSelector(switchChainModalStore, (state) => state.context.onClose);
