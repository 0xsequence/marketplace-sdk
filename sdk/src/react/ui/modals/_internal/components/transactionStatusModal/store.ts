import type { QueryKey } from '@tanstack/react-query';
import { createStore } from '@xstate/store';
import { useSelector } from '@xstate/store/react';
import type { Hex } from 'viem';
import type { Price } from '../../../../../../types';
import type { TransactionType } from '../../../../../_internal/types';
import type { ModalCallbacks } from '../../types';
import type { ShowTransactionStatusModalArgs } from '.';

export type ConfirmationStatus = {
	isConfirming: boolean;
	isConfirmed: boolean;
	isFailed: boolean;
	isTimeout: boolean;
};

export type TransactionStatus = 'PENDING' | 'SUCCESS' | 'FAILED' | 'TIMEOUT';

interface TransactionStatusModalContext {
	isOpen: boolean;
	hash: Hex | undefined;
	orderId: string | undefined;
	status: TransactionStatus;
	transactionType: TransactionType | undefined;
	price: Price | undefined;
	collectionAddress: Hex;
	chainId: number;
	collectibleId: string;
	callbacks?: ModalCallbacks;
	queriesToInvalidate?: QueryKey[];
}

const initialContext: TransactionStatusModalContext = {
	isOpen: false,
	hash: undefined,
	orderId: undefined,
	status: 'PENDING',
	transactionType: undefined,
	price: undefined,
	collectionAddress: '' as Hex,
	chainId: 0,
	collectibleId: '',
	callbacks: undefined,
	queriesToInvalidate: [],
};

export const transactionStatusModalStore = createStore({
	context: initialContext,
	on: {
		open: (
			context,
			event: Omit<ShowTransactionStatusModalArgs, 'type'> & {
				transactionType: TransactionType;
			},
		) => ({
			...context,
			isOpen: true,
			hash: event.hash,
			orderId: event.orderId,
			price: event.price,
			collectionAddress: event.collectionAddress,
			chainId: event.chainId,
			collectibleId: event.collectibleId,
			transactionType: event.transactionType,
			callbacks: event.callbacks,
			queriesToInvalidate: event.queriesToInvalidate,
			status: 'PENDING' as TransactionStatus,
		}),

		close: () => ({
			...initialContext,
		}),

		updateStatus: (context, event: { status: TransactionStatus }) => ({
			...context,
			status: event.status,
		}),
	},
});

// Selector hooks
export const useIsOpen = () =>
	useSelector(transactionStatusModalStore, (state) => state.context.isOpen);

export const useTransactionModalState = () =>
	useSelector(transactionStatusModalStore, (state) => state.context);

export const useTransactionHash = () =>
	useSelector(transactionStatusModalStore, (state) => state.context.hash);

export const useTransactionType = () =>
	useSelector(
		transactionStatusModalStore,
		(state) => state.context.transactionType,
	);

export const useTransactionDetails = () =>
	useSelector(transactionStatusModalStore, (state) => ({
		hash: state.context.hash,
		orderId: state.context.orderId,
		type: state.context.transactionType,
		status: state.context.status,
		chainId: state.context.chainId,
	}));

export const useCollectibleInfo = () =>
	useSelector(transactionStatusModalStore, (state) => ({
		collectionAddress: state.context.collectionAddress,
		collectibleId: state.context.collectibleId,
		chainId: state.context.chainId,
		price: state.context.price,
	}));
