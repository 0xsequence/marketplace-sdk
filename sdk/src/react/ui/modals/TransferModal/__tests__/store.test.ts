import type { Address } from 'viem';
import { beforeEach, describe, expect, it } from 'vitest';
import type { ShowTransferModalArgs } from '../index';
import { transferModalStore } from '../store';

describe('TransferModal Store', () => {
	beforeEach(() => {
		// Reset store to initial state before each test
		transferModalStore.send({ type: 'close' });
	});

	describe('Initial State', () => {
		it('should have correct initial state', () => {
			const state = transferModalStore.getSnapshot();

			expect(state.context).toMatchSnapshot('initial-store-state');
			expect(state.context.isOpen).toBe(false);
			expect(state.context.view).toBe('enterReceiverAddress');
			expect(state.context.state.receiverAddress).toBe('');
			expect(state.context.state.quantity).toBe('1');
			expect(state.context.state.transferIsBeingProcessed).toBe(false);
			expect(state.context.hash).toBeUndefined();
		});
	});

	describe('Open Action', () => {
		const mockOpenArgs: ShowTransferModalArgs = {
			collectionAddress: '0x742d35Cc6634C0532925a3b8D4C9db96' as Address,
			collectibleId: '123',
			chainId: 1,
			callbacks: {
				onSuccess: () => {},
				onError: () => {},
			},
		};

		it('should open modal with provided arguments', () => {
			transferModalStore.send({ type: 'open', ...mockOpenArgs });

			const state = transferModalStore.getSnapshot();

			expect(state.context.isOpen).toBe(true);
			expect(state.context.state.collectionAddress).toBe(
				mockOpenArgs.collectionAddress,
			);
			expect(state.context.state.collectibleId).toBe(
				mockOpenArgs.collectibleId,
			);
			expect(state.context.state.chainId).toBe(mockOpenArgs.chainId);
			expect(state.context.state.callbacks).toBe(mockOpenArgs.callbacks);
		});

		it('should preserve existing state when opening', () => {
			// Set some state first
			transferModalStore.send({ type: 'setReceiverAddress', address: '0x123' });
			transferModalStore.send({ type: 'setQuantity', quantity: '5' });

			transferModalStore.send({ type: 'open', ...mockOpenArgs });

			const state = transferModalStore.getSnapshot();

			expect(state.context.isOpen).toBe(true);
			expect(state.context.state.receiverAddress).toBe('0x123');
			expect(state.context.state.quantity).toBe('5');
		});

		it('should create snapshot of opened state', () => {
			transferModalStore.send({ type: 'open', ...mockOpenArgs });

			const state = transferModalStore.getSnapshot();
			expect(state.context).toMatchSnapshot('store-after-open');
		});
	});

	describe('Close Action', () => {
		it('should reset to initial state when closed', () => {
			const mockOpenArgs: ShowTransferModalArgs = {
				collectionAddress: '0x742d35Cc6634C0532925a3b8D4C9db96' as Address,
				collectibleId: '123',
				chainId: 1,
			};

			// Open and modify state
			transferModalStore.send({ type: 'open', ...mockOpenArgs });
			transferModalStore.send({ type: 'setReceiverAddress', address: '0x456' });
			transferModalStore.send({ type: 'setQuantity', quantity: '10' });
			transferModalStore.send({
				type: 'setView',
				view: 'followWalletInstructions',
			});

			// Close modal
			transferModalStore.send({ type: 'close' });

			const state = transferModalStore.getSnapshot();

			expect(state.context.isOpen).toBe(false);
			expect(state.context.view).toBe('enterReceiverAddress');
			expect(state.context.state.receiverAddress).toBe('');
			expect(state.context.state.quantity).toBe('1');
			expect(state.context.state.transferIsBeingProcessed).toBe(false);
		});
	});

	describe('SetView Action', () => {
		it('should update view state', () => {
			transferModalStore.send({
				type: 'setView',
				view: 'followWalletInstructions',
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context.view).toBe('followWalletInstructions');
		});

		it('should handle undefined view', () => {
			transferModalStore.send({ type: 'setView', view: undefined });

			const state = transferModalStore.getSnapshot();
			expect(state.context.view).toBeUndefined();
		});
	});

	describe('SetHash Action', () => {
		it('should set transaction hash', () => {
			const hash = '0xabcdef123456789' as `0x${string}`;
			transferModalStore.send({ type: 'setHash', hash });

			const state = transferModalStore.getSnapshot();
			expect(state.context.hash).toBe(hash);
		});

		it('should handle undefined hash', () => {
			transferModalStore.send({ type: 'setHash', hash: undefined });

			const state = transferModalStore.getSnapshot();
			expect(state.context.hash).toBeUndefined();
		});
	});

	describe('SetReceiverAddress Action', () => {
		it('should update receiver address', () => {
			const address = '0x742d35Cc6634C0532925a3b8D4C9db96';
			transferModalStore.send({ type: 'setReceiverAddress', address });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.receiverAddress).toBe(address);
		});

		it('should handle empty address', () => {
			transferModalStore.send({ type: 'setReceiverAddress', address: '' });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.receiverAddress).toBe('');
		});
	});

	describe('SetQuantity Action', () => {
		it('should update quantity', () => {
			transferModalStore.send({ type: 'setQuantity', quantity: '5' });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.quantity).toBe('5');
		});

		it('should handle zero quantity', () => {
			transferModalStore.send({ type: 'setQuantity', quantity: '0' });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.quantity).toBe('0');
		});

		it('should handle string quantity', () => {
			transferModalStore.send({ type: 'setQuantity', quantity: '100' });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.quantity).toBe('100');
		});
	});

	describe('SetTransferIsBeingProcessed Action', () => {
		it('should set processing state to true', () => {
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: true,
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.transferIsBeingProcessed).toBe(true);
		});

		it('should set processing state to false', () => {
			// First set to true
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: true,
			});

			// Then set to false
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: false,
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.transferIsBeingProcessed).toBe(false);
		});

		it('should create snapshot of processing state', () => {
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: true,
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context).toMatchSnapshot('store-processing-transfer');
		});
	});

	describe('UpdateState Action', () => {
		it('should partially update state', () => {
			const partialUpdate = {
				receiverAddress: '0x123',
				quantity: '10',
				chainId: 137,
			};

			transferModalStore.send({ type: 'updateState', ...partialUpdate });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.receiverAddress).toBe('0x123');
			expect(state.context.state.quantity).toBe('10');
			expect(state.context.state.chainId).toBe(137);
			// Other fields should remain unchanged
			expect(state.context.state.transferIsBeingProcessed).toBe(false);
		});

		it('should handle callbacks update', () => {
			const callbacks = {
				onSuccess: () => console.log('success'),
				onError: () => console.log('error'),
			};

			transferModalStore.send({ type: 'updateState', callbacks });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.callbacks).toBe(callbacks);
		});

		it('should merge with existing state', () => {
			// Set initial state
			transferModalStore.send({ type: 'setReceiverAddress', address: '0x456' });
			transferModalStore.send({ type: 'setQuantity', quantity: '5' });

			// Partial update
			transferModalStore.send({ type: 'updateState', chainId: 1 });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.receiverAddress).toBe('0x456');
			expect(state.context.state.quantity).toBe('5');
			expect(state.context.state.chainId).toBe(1);
		});
	});

	describe('Selector Hooks Integration', () => {
		it('should work with useIsOpen selector', () => {
			// This test verifies the store works with the selector hooks
			// The actual hook testing will be done in component tests

			const initialState = transferModalStore.getSnapshot();
			expect(initialState.context.isOpen).toBe(false);

			transferModalStore.send({
				type: 'open',
				collectionAddress: '0x123' as Address,
				collectibleId: '1',
				chainId: 1,
			});

			const openState = transferModalStore.getSnapshot();
			expect(openState.context.isOpen).toBe(true);
		});

		it('should work with useModalState selector', () => {
			const address = '0x742d35Cc6634C0532925a3b8D4C9db96';
			transferModalStore.send({ type: 'setReceiverAddress', address });

			const state = transferModalStore.getSnapshot();
			expect(state.context.state.receiverAddress).toBe(address);
		});

		it('should work with useView selector', () => {
			transferModalStore.send({
				type: 'setView',
				view: 'followWalletInstructions',
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context.view).toBe('followWalletInstructions');
		});

		it('should work with useHash selector', () => {
			const hash = '0xabcdef' as `0x${string}`;
			transferModalStore.send({ type: 'setHash', hash });

			const state = transferModalStore.getSnapshot();
			expect(state.context.hash).toBe(hash);
		});
	});

	describe('Complex State Transitions', () => {
		it('should handle complete transfer flow state changes', () => {
			const mockArgs: ShowTransferModalArgs = {
				collectionAddress: '0x742d35Cc6634C0532925a3b8D4C9db96' as Address,
				collectibleId: '123',
				chainId: 1,
			};

			// 1. Open modal
			transferModalStore.send({ type: 'open', ...mockArgs });
			expect(transferModalStore.getSnapshot().context.isOpen).toBe(true);

			// 2. Set receiver address
			transferModalStore.send({ type: 'setReceiverAddress', address: '0x456' });
			expect(
				transferModalStore.getSnapshot().context.state.receiverAddress,
			).toBe('0x456');

			// 3. Set quantity for ERC1155
			transferModalStore.send({ type: 'setQuantity', quantity: '5' });
			expect(transferModalStore.getSnapshot().context.state.quantity).toBe('5');

			// 4. Start processing
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: true,
			});
			expect(
				transferModalStore.getSnapshot().context.state.transferIsBeingProcessed,
			).toBe(true);

			// 5. Switch to wallet instructions view
			transferModalStore.send({
				type: 'setView',
				view: 'followWalletInstructions',
			});
			expect(transferModalStore.getSnapshot().context.view).toBe(
				'followWalletInstructions',
			);

			// 6. Set transaction hash
			const hash = '0xabcdef123456789' as `0x${string}`;
			transferModalStore.send({ type: 'setHash', hash });
			expect(transferModalStore.getSnapshot().context.hash).toBe(hash);

			// 7. Complete transfer (close modal)
			transferModalStore.send({ type: 'close' });
			const finalState = transferModalStore.getSnapshot();
			expect(finalState.context.isOpen).toBe(false);
			expect(finalState.context.view).toBe('enterReceiverAddress');
			expect(finalState.context.state.transferIsBeingProcessed).toBe(false);
		});

		it('should create snapshot of complex state', () => {
			const mockArgs: ShowTransferModalArgs = {
				collectionAddress: '0x742d35Cc6634C0532925a3b8D4C9db96' as Address,
				collectibleId: '123',
				chainId: 1,
				callbacks: {
					onSuccess: () => {},
					onError: () => {},
				},
			};

			transferModalStore.send({ type: 'open', ...mockArgs });
			transferModalStore.send({ type: 'setReceiverAddress', address: '0x456' });
			transferModalStore.send({ type: 'setQuantity', quantity: '10' });
			transferModalStore.send({
				type: 'setTransferIsBeingProcessed',
				isProcessing: true,
			});

			const state = transferModalStore.getSnapshot();
			expect(state.context).toMatchSnapshot('store-complex-state');
		});
	});
});
