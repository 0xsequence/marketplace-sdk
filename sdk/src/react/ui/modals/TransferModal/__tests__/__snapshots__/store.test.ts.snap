// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransferModal Store > Complex State Transitions > should create snapshot of complex state > store-complex-state 1`] = `
{
  "hash": undefined,
  "isOpen": true,
  "state": {
    "callbacks": {
      "onError": [Function],
      "onSuccess": [Function],
    },
    "chainId": 1,
    "collectibleId": "123",
    "collectionAddress": "0x742d35Cc6634C0532925a3b8D4C9db96",
    "quantity": "10",
    "receiverAddress": "0x456",
    "transferIsBeingProcessed": true,
  },
  "view": "enterReceiverAddress",
}
`;

exports[`TransferModal Store > Initial State > should have correct initial state > initial-store-state 1`] = `
{
  "hash": undefined,
  "isOpen": false,
  "state": {
    "chainId": 0,
    "collectibleId": "",
    "collectionAddress": "0x",
    "quantity": "1",
    "receiverAddress": "",
    "transferIsBeingProcessed": false,
  },
  "view": "enterReceiverAddress",
}
`;

exports[`TransferModal Store > Open Action > should create snapshot of opened state > store-after-open 1`] = `
{
  "hash": undefined,
  "isOpen": true,
  "state": {
    "callbacks": {
      "onError": [Function],
      "onSuccess": [Function],
    },
    "chainId": 1,
    "collectibleId": "123",
    "collectionAddress": "0x742d35Cc6634C0532925a3b8D4C9db96",
    "quantity": "1",
    "receiverAddress": "",
    "transferIsBeingProcessed": false,
  },
  "view": "enterReceiverAddress",
}
`;

exports[`TransferModal Store > SetTransferIsBeingProcessed Action > should create snapshot of processing state > store-processing-transfer 1`] = `
{
  "hash": undefined,
  "isOpen": false,
  "state": {
    "chainId": 0,
    "collectibleId": "",
    "collectionAddress": "0x",
    "quantity": "1",
    "receiverAddress": "",
    "transferIsBeingProcessed": true,
  },
  "view": "enterReceiverAddress",
}
`;
